internal/services/netapp/validate/volume_group_sap_hana_volumes_validation.go

```diff
@@ -0,0 +1,159 @@
+package validate
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/go-azure-helpers/lang/pointer"
+	"github.com/hashicorp/go-azure-sdk/resource-manager/netapp/2022-05-01/volumegroups"
+	"github.com/hashicorp/terraform-provider-azurerm/utils"
+)
+
+type VolumeSpecNameSapHana string
+
+const (
+	VolumeSpecNameSapHanaData       VolumeSpecNameSapHana = "data"
+	VolumeSpecNameSapHanaLog        VolumeSpecNameSapHana = "log"
+	VolumeSpecNameSapHanaShared     VolumeSpecNameSapHana = "shared"
+	VolumeSpecNameSapHanaDataBackup VolumeSpecNameSapHana = "data-backup"
+	VolumeSpecNameSapHanaLogBackup  VolumeSpecNameSapHana = "log-backup"
+)
+
+func PossibleValuesForVolumeSpecNameSapHana() []string {
+	return []string{
+		string(VolumeSpecNameSapHanaData),
+		string(VolumeSpecNameSapHanaLog),
+		string(VolumeSpecNameSapHanaShared),
+		string(VolumeSpecNameSapHanaDataBackup),
+		string(VolumeSpecNameSapHanaLogBackup),
+	}
+}
+
+func RequiredVolumesForSAPHANA() []string {
+	return []string{
+		string(VolumeSpecNameSapHanaData),
+		string(VolumeSpecNameSapHanaLog),
+	}
+}
+
+type ProtocolType string
+
+const (
+	ProtocolTypeNfsV41 ProtocolType = "NFSv4.1"
+	ProtocolTypeNfsV3  ProtocolType = "NFSv3"
+	ProtocolTypeCifs   ProtocolType = "CIFS"
+)
+
+func PossibleValuesForProtocolType() []string {
+	return []string{
+		string(ProtocolTypeNfsV41),
+		string(ProtocolTypeNfsV3),
+		string(ProtocolTypeCifs),
+	}
+}
+
+func PossibleValuesForProtocolTypeVolumeGroupSapHana() []string {
+	return []string{
+		string(ProtocolTypeNfsV41),
+		string(ProtocolTypeNfsV3),
+	}
+}
+
+// Diverging from the SDK volumegroups.SecurityStyle since it is defined as lower case
+// but the backend changes it to Pascal case on GET. Please refer to https://github.com/Azure/azure-sdk-for-go/issues/14684
+type SecurityStyle string
+
+const (
+	SecurityStyleUnix SecurityStyle = "Unix"
+	SecurityStyleNtfs SecurityStyle = "Ntfs"
+)
+
+func PossibleValuesForSecurityStyle() []string {
+	return []string{
+		string(SecurityStyleUnix),
+	}
+}
+
+func ValidateNetAppVolumeGroupSAPHanaVolumes(volumeList *[]volumegroups.VolumeGroupVolumeProperties) []error {
+	errors := make([]error, 0)
+	volumeSpecRepeatCount := make(map[string]int)
+	applicationType := string(volumegroups.ApplicationTypeSAPNegativeHANA)
+
+	// Validating each volume
+	for _, volume := range pointer.From(volumeList) {
+		protocolTypeList := pointer.From(volume.Properties.ProtocolTypes)
+		protocolType := ""
+		if len(protocolTypeList) > 0 {
+			protocolType = protocolTypeList[0]
```

Comment:
This might not start with an `nfs` protocol when multi protocol support is added. Maybe we could loop through it until we find an `nfs` protocol and then set protcolType to that?

Reply:
Included extra validations for protocolTypes to handle this without giving away hypothetical roadmaps: `ValidateMultiProtocolFails`, `ValidateNoProtocolFails`, `ValidateInvalidProtocol`, `ValidateCIFSInvalidProtocolForSAPHana`.

```
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateCorrectSettingsAllVolumes
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateMinimumVolumes
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateRequiredVolumeSpecs
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateLessThanMinimumVolumes
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateMultiProtocolFails
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateNoProtocolFails
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateInvalidProtocol
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateCIFSInvalidProtocolForSAPHana
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateNoNfsVersionThreeOnDataLogAndSharedVolumes
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateNoPPGBackupVolumes
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateRequiredPpgForNonBackupVolumes
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateVolumeSpecCantRepeat
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateEndpointDstNotEnabledOnLogVolume
=== RUN   TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateSnapshotPolicyNotEnabledOnEndpointDstVolume
--- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes (0.01s)
    --- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateCorrectSettingsAllVolumes (0.00s)
    --- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateMinimumVolumes (0.00s)
    --- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateRequiredVolumeSpecs (0.00s)
    --- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateLessThanMinimumVolumes (0.00s)
    --- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateMultiProtocolFails (0.00s)
    --- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateNoProtocolFails (0.00s)
    --- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateInvalidProtocol (0.00s)
    --- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateCIFSInvalidProtocolForSAPHana (0.00s)
    --- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateNoNfsVersionThreeOnDataLogAndSharedVolumes (0.00s)
    --- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateNoPPGBackupVolumes (0.00s)
    --- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateRequiredPpgForNonBackupVolumes (0.00s)
    --- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateVolumeSpecCantRepeat (0.00s)
    --- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateEndpointDstNotEnabledOnLogVolume (0.00s)
    --- PASS: TestValidateNetAppVolumeGroupSAPHanaVolumes/ValidateSnapshotPolicyNotEnabledOnEndpointDstVolume (0.00s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/netapp/validate       0.017s
```